### PR TITLE
Add some error handling to OTP secret migration

### PR DESCRIPTION
**Pull request by ClearlyClaire**
Attempt to address https://github.com/mastodon/mastodon/pull/28325#issuecomment-2116365991 by aborting with a more understandable error spelling out that decrypting OTP secrets failed and that it is most likely due to a past change to `OTP_SECRET`.

Also allows ignoring such failures through an environment variable.

> == 20240307180905 MigrateDeviseTwoFactorSecrets: migrating ====================
> 
> ERROR: Unable to decrypt OTP secret for user `@shanon0`.
> 
> This is most likely because you have changed the value of `OTP_SECRET` at some point in
time after `@shanon0` configured 2FA.
> 
> In this case, their OTP secret had already been lost with the change to `OTP_SECRET`, and
proceeding with this migration will not make the situation worse.
> 
> Please double-check that you have not accidentally changed `OTP_SECRET` just for this
migration, and re-run the migration with `MIGRATION_IGNORE_INVALID_OTP_SECRET=true`.
> 
> Migration aborted.

### CODE REVIEW

File: db/post_migrate/20240307180905_migrate_devise_two_factor_secrets.rb [hunk 1]
```diff
@@ -36,4 +42,22 @@ def down
   def users_with_otp_enabled
     MigrationUser.where(otp_required_for_login: true, otp_secret: nil)
   end
+
+  def abort_with_decryption_error(user)
+    abort <<~MESSAGE
+
+      ERROR: Unable to decrypt OTP secret for user @#{user.account.username}.
```
Comment (mjankowski) [inline db/post_migrate/20240307180905_migrate_devise_two_factor_secrets.rb:49]: In this spot, isn't `user` an instance of `MigrationUser`, and if so, does `#account` exist?

Comment (ClearlyClaire) [reply db/post_migrate/20240307180905_migrate_devise_two_factor_secrets.rb:49]: Oof, you're right! Good catch!

**Review by mjankowski — APPROVED** [review]
I have a slight fear that this message will be more confusing to people than if we just silently continued the migration when errors occurred, but I can't think of a more clear way to write it, given what we're trying to say.

I guess we also don't really have a way of knowing what portion of people will ever see this message, and of those how many actually have any users who have not abandoned their accounts but unknowingly have broken 2FA. Must be a small number.

All that said, I think this is a net improvement over the current scenario and it handles the error condition at least.

**Review by renchap — APPROVED** [review]

**Comment by itbeard** [comment]
Hey!
I got this message while upgrading from 4.2.13 to 4.3.0.

```
Oct 22 00:48:48 ERROR: Unable to decrypt OTP secret for user 1.

Oct 22 00:48:48 2024-10-21T22:48:48.000Z

Oct 22 00:48:48 This is most likely because you have changed the value of `OTP_SECRET` at some point in

Oct 22 00:48:48 time after the user configured 2FA.

Oct 22 00:48:48 2024-10-21T22:48:48.000Z

Oct 22 00:48:48 In this case, their OTP secret had already been lost with the change to `OTP_SECRET`, and

Oct 22 00:48:48 proceeding with this migration will not make the situation worse.

Oct 22 00:48:48 2024-10-21T22:48:48.000Z

Oct 22 00:48:48 Please double-check that you have not accidentally changed `OTP_SECRET` just for this

Oct 22 00:48:48 migration, and re-run the migration with `MIGRATION_IGNORE_INVALID_OTP_SECRET=true`.

Oct 22 00:48:48 2024-10-21T22:48:48.000Z

Oct 22 00:48:48 Migration aborted.
```

I checked the OTP_SECRET value - it is the same before and after the update. I'm using Claudrone for hosting my Masto instance btw. Any ideas on how to fix it?

**Comment by ClearlyClaire** [comment]
The `OTP_SECRET` value change could have happened at any point in your instance's lifetime.

A common source of `OTP_SECRET` changing over a instance's lifetime is it being manually copied from one server to the next and accidentally truncated in the process.

If you don't have a backup of the old value, the information is already lost and the affected user is already unable to use TOTP, so running the migration again with `MIGRATION_IGNORE_INVALID_OTP_SECRET=true` should not break anything more, and will allow you to proceed. Be sure to set the encryption secrets once and not change them in the future.

**Pull request merged**